### PR TITLE
Add a config not to shuffle merged dataset

### PR DESCRIPTION
**Pull request by seungduk-yanolja**
Add a config not to shuffle merged dataset

# Description

Added a config named `not_shuffle_merged_datasets`, which I have been using in my fork for a long time :)

## Motivation and Context

When training a model to expand its vocab with non-English tokens, I usually start with parallel corpora and then train it on web-crawled or something suitable for pre-training.
It is better giving the user to have an option not to shuffle the merged datasets anyway.

## How has this been tested?

This config has been used in my fork for a long time and I verified that it works by seeing the loss graph.

## Screenshots (if appropriate)

N/A

## Types of changes

New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by seungduk-yanolja** [comment]
Resolves #1393

File: src/axolotl/utils/data.py [hunk 1]
```diff
@@ -819,7 +822,11 @@ def wrap_pretraining_dataset(
     else:
         encode = functools.partial(encode_pretraining, tokenizer, max_tokens)
 
-    dataset = dataset.shuffle(seed=seed, buffer_size=buffer_size)
+    if cfg.not_shuffle_merged_datasets:
+        LOG.info("NOT shuffling merged pretraining datasets")
+    else:
+        dataset = dataset.shuffle(seed=seed, buffer_size=buffer_size)
```
Comment (seungduk-yanolja) [inline src/axolotl/utils/data.py:826]: I am unsure if this is intended to shuffle the pre-training dataset (it is a single dataset) within the buffer size?

File: README.md [hunk 2]
```diff
@@ -662,6 +662,10 @@ datasets:
       # For `completion` datsets only, uses the provided field instead of `text` column
       field:
 
+# If true, the datasets will not be shuffled and will keep their original order in `datasets`.
+# The same applies to the `test_datasets` option and the `pretraining_dataset` option. Default is false.
+not_shuffle_merged_datasets: false
```
Comment (winglian) [inline README.md]: I think inverted configuration variables are harder to understand. It's probably easier to use `shuffle_merged_datasets: false` where the default is true.

File: src/axolotl/utils/config/models/input/v0_4_1/__init__.py [hunk 3]
```diff
@@ -411,6 +411,7 @@ class Config:
 
     datasets: Optional[conlist(Union[SFTDataset, DPODataset], min_length=1)] = None  # type: ignore
     test_datasets: Optional[conlist(Union[SFTDataset, DPODataset], min_length=1)] = None  # type: ignore
+    not_shuffle_merged_datasets: Optional[bool] = None
```
Comment (winglian) [inline src/axolotl/utils/config/models/input/v0_4_1/__init__.py]:
```suggestion
    shuffle_merged_datasets: Optional[bool] = True
```

File: src/axolotl/utils/data.py [hunk 4]
```diff
@@ -415,8 +415,11 @@ def for_d_in_datasets(dataset_configs):
         dataset = concatenate_datasets(datasets)
 
         if len(datasets) > 1:
-            LOG.info("shuffle merged datasets")
-            dataset = dataset.shuffle(seed=seed)
+            if cfg.not_shuffle_merged_datasets:
```
Comment (winglian) [inline src/axolotl/utils/data.py]:
```suggestion
            if cfg.shuffle_merged_datasets is False:
```

File: src/axolotl/utils/data.py [hunk 5]
```diff
@@ -415,8 +415,11 @@ def for_d_in_datasets(dataset_configs):
...
+                LOG.info("NOT shuffling merged datasets")
+            else:
+                LOG.info("shuffle merged datasets")
```
Comment (winglian) [inline src/axolotl/utils/data.py]:
```suggestion
                LOG.debug("NOT shuffling merged datasets")
            else:
                LOG.debug("shuffle merged datasets")
```

File: src/axolotl/utils/data.py (hunk 1 above)
Comment (winglian) [inline src/axolotl/utils/data.py]:
```suggestion
    if cfg.shuffle_merged_datasets is False:
        LOG.debug("NOT shuffling merged pretraining datasets")
    else:
        dataset = dataset.shuffle(seed=seed, buffer_size=buffer_size)
```

**Review by winglian — DISMISSED** [review]

File: README.md (hunk 2 above)
Comment (seungduk-yanolja) [reply README.md]: Done

File: src/axolotl/utils/data.py (hunk 4 above)
Comment (seungduk-yanolja) [reply src/axolotl/utils/data.py]: inverted the condition

File: src/axolotl/utils/data.py (hunk 5 above)
Comment (seungduk-yanolja) [reply src/axolotl/utils/data.py]: done

File: src/axolotl/utils/data.py (hunk 1 above)
Comment (seungduk-yanolja) [reply src/axolotl/utils/data.py]: done
inverted the condition

**Comment by NanoCode012** [comment]
I am a bit curious. This only disables shuffling when merging datasets. It is still intended to shuffle within dataset?

**Comment by seungduk-yanolja** [comment]
Does this PR require some more changes to be merged? PTAL

**Review by NanoCode012 — APPROVED** [review]

**Comment by Gryphe** [comment]
Can anyone confirm that the order in which the datasets are processed is actually reversed, with the last YAML dataset entry being processed first? 

I was running some experiments with this new option and the spikes in my grad_norm seem to indicate that my "most difficult" dataset, which is the last entry in my YAML, is actually processed first. I'm running more experiments as we speak, currently one in which the dataset order has been reversed.
